subsys/fs/nvs/nvs.c

```diff
@@ -6,104 +6,492 @@
+	}
+	if (padding) {
+		/* add reminder of padding to buf */
+		memset(buf + byte_cnt, 0xff, padding);
+		byte_cnt += padding;
+	}
+	/* add ilenstr at the end */
+	memcpy(buf+byte_cnt, ilenstr, length_cnt);
+	byte_cnt += length_cnt;
+	rc = _nvs_flash_write_ver(fs, offset, buf, byte_cnt);
+	if (rc) {
+		/* flash write or verify error */
+		goto err;
+	}
+	offset += byte_cnt;
+	fs->write_location = offset;
+
+	if (fs->write_location >= fs_size) {
+		fs->write_location -= fs_size;
+	}
+
+	return 0;
+
+err:
+	/* lock the filesystem */
+	fs->locked = true;
+	SYS_LOG_ERR("File system is locked");
+	return rc;
+}
+/* read the length and id from flash at location offset (at start of item),
+ * returns 0xffff in len and id if the read length fields are ff and ff
+ */
+int _nvs_flash_rd_len_id_start(struct nvs_fs *fs, off_t offset, u16_t *id,
+			       size_t *len)
+{
+	int rc;
+	u8_t lenstr[sizeof(u16_t)];
+	size_t len_size;
+	off_t addr;
+
+	addr = offset;
+	rc = _nvs_flash_read(fs, addr, lenstr, sizeof(u16_t));
+	if (rc) {
+		return rc;
+	}
+	len_size = _nvs_get_length(lenstr, len);
+	if (len_size == 0xffff) {
+		*id = NVS_ID_EMPTY;
+		*len = 0xffff;
+		return 0;
+	}
+	_nvs_addr_advance(fs, &addr, len_size);
+	rc = _nvs_flash_read(fs, addr, id, sizeof(u16_t));
+	return rc;
+}
+
+/* read the length and id from flash just before location offset (at end of
+ * item), returns 0xffff in len and id if the read flash fields are ff.
+ */
+int _nvs_flash_rd_len_id_end(struct nvs_fs *fs, off_t offset, u16_t *id,
+			     size_t *len)
+{
+	int rc;
+	size_t len_size;
+	u8_t lenstr[sizeof(u16_t)], ilenstr[sizeof(u16_t)];
+	off_t addr;
+
+	addr = offset;
+	_nvs_addr_decrease(fs, &addr, sizeof(u16_t));
+	rc = _nvs_flash_read(fs, addr, ilenstr, sizeof(u16_t));
+	if (rc) {
+		return rc;
+	}
+	lenstr[0] = ilenstr[1];
+	lenstr[1] = ilenstr[0];
+	len_size = _nvs_get_length(lenstr, len);
+	if (rc == 0xffff) {
+		*id = NVS_ID_EMPTY;
+		*len = 0xffff;
+		return 0;
+	}
+	addr = offset;
+	_nvs_addr_decrease(fs, &addr, _nvs_flash_size(fs, *len));
+	_nvs_addr_advance(fs, &addr, len_size);
+	rc = _nvs_flash_read(fs, addr, id, sizeof(u16_t));
+	return rc;
+}
+
+/* read the data from flash, the offset is the start of the entry
+ * first read the length from flash as partial reading of the data can be
+ * requested
+ */
+ssize_t _nvs_flash_read_data(struct nvs_fs *fs, off_t offset, void *data,
+			     size_t len)
+{
+	int rc;
+	u8_t lenstr[sizeof(u16_t)];
+	size_t len_size;
+	size_t len_flash;
+	off_t addr;
+
+	addr = offset;
+	rc = _nvs_flash_read(fs, addr, lenstr, sizeof(u16_t));
+	if (rc) {
+		return rc;
+	}
+	len_size = _nvs_get_length(lenstr, &len_flash);
+
+	if (len_size == 1) {
+		_nvs_addr_advance(fs, &addr, sizeof(u8_t) + sizeof(u16_t));
+	}
+	if (len_size == 2) {
+		_nvs_addr_advance(fs, &addr, sizeof(u16_t) + sizeof(u16_t));
+	}
+
+	/* if len_size = 0xffff just read from the address given as all should
+	 * be ff
+	 */
+	rc = _nvs_flash_read(fs, addr, data, len);
```

Comment:
There you probably want to use `min(len, len_flash)`, to make sure the data from the next entry, or even from outside the NVS area is read. This could be a security issue when combined with missing input validation from the caller. I actually wonder if `_nvs_flash_read` should also limit the access to the NVS area only.